extensions/amp-graphiq/0.1/amp-graphiq.js

```diff
@@ -0,0 +1,181 @@
+ * </amp-graphiq>
+ * </code>
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {setStyles} from '../../../src/style';
+import {removeElement} from '../../../src/dom';
+import {user} from '../../../src/log';
+import {tryParseJson} from '../../../src/json';
+import {isObject} from '../../../src/types';
+import {listen} from '../../../src/event-helper';
+
+class AmpGraphiq extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.iframePromise_ = null;
+
+    /** @private {!string} */
+    this.widgetId_ = '';
+
+    /** @private {!string} */
+    this.href_ = '';
+
+    /** @private {!string} */
+    this.initialHeight_ = '';
+
+    /** @private {!string} */
+    this.initialWidth_ = '';
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+  }
+ /**
+  * @param {boolean=} opt_onLayout
+  * @override
+  */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url('https://w.graphiq.com', opt_onLayout);
+  }
+
+  /** @override */
+  renderOutsideViewport() {
+    return false;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.widgetId_ = user().assert(
+        this.element.getAttribute('data-widget-id'),
+        'The data-widget-id attribute is required for <amp-graphiq> %s',
+        this.element);
+    this.href_ = this.element.getAttribute('data-href') ||
+        'https://www.graphiq.com/vlp/' + this.widgetId_;
+    this.initialHeight_ = this.element.getAttribute('height');
+    this.initialWidth_ = this.element.getAttribute('width');
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    this.iframe_ = iframe;
+
+    this.unlistenMessage_ = listen(
+      this.win,
+      'message',
+      this.handleGraphiqMessages_.bind(this)
+    );
+
+    iframe.setAttribute('frameborder', '0');
+    iframe.setAttribute('allowtransparency', 'true');
+    //Add title to the iframe for better accessibility.
+    iframe.setAttribute('title', 'Graphiq: ' +
+        this.element.getAttribute('alt') || '');
+    if (this.initialHeight_) {
+      iframe.setAttribute('height', this.initialHeight_);
+    }
+    if (this.initialWidth_) {
+      iframe.setAttribute('width', this.initialWidth_);
+    }
+    iframe.src = 'https://w.graphiq.com/w/' + this.widgetId_ +
+        '?data-width=' + encodeURIComponent(this.initialWidth_) +
+        '&data-height=' + encodeURIComponent(this.initialHeight_) +
+        '&data-href=' + encodeURIComponent(this.href_);
+
+    this.applyFillContent(iframe);
+    this.element.appendChild(iframe);
+    setStyles(iframe, {
+      'opacity': 0,
+    });
+    return this.iframePromise_ = this.loadPromise(iframe).then(() => {
```

Comment:
Not sure - this is a pattern I'd seen in a couple of other components (eg `amp-instagram`) - my assumption was that you had to ensure the iframe had completed loading before calling `this.getVsync().mutate()`, but I didn't closely to see if that was in fact true.  Upon further investigation I only see this done in 3 or 4 components, so not sure.  I'll remove it if you think it's unnecessary.